internal/idle/idle.go

```diff
@@ -93,52 +76,51 @@ type manager struct {
 type ManagerOptions struct {
 	Enforcer Enforcer
 	Timeout  time.Duration
-	Logger   grpclog.LoggerV2
 }
 
 // NewManager creates a new idleness manager implementation for the
-// given idle timeout.
-func NewManager(opts ManagerOptions) Manager {
-	if opts.Timeout == 0 {
-		return noopManager{}
+// given idle timeout.  It begins in idle mode.
+func NewManager(opts ManagerOptions) *Manager {
+	return &Manager{
+		enforcer:         opts.Enforcer,
+		timeout:          opts.Timeout,
+		actuallyIdle:     true,
+		activeCallsCount: -math.MaxInt32,
 	}
-
-	m := &manager{
-		enforcer: opts.Enforcer,
-		timeout:  int64(opts.Timeout),
-		logger:   opts.Logger,
-	}
-	m.timer = timeAfterFunc(opts.Timeout, m.handleIdleTimeout)
-	return m
 }
 
 // resetIdleTimer resets the idle timer to the given duration. This method
 // should only be called from the timer callback.
-func (m *manager) resetIdleTimer(d time.Duration) {
-	m.idleMu.Lock()
-	defer m.idleMu.Unlock()
-
-	if m.timer == nil {
-		// Only close sets timer to nil. We are done.
+func (m *Manager) resetIdleTimerLocked(d time.Duration) {
+	if m.isClosed() || m.timeout == 0 {
```

Comment:
Should we also return early with a check for `m.timeout == 0` in `OnCallXxx` and save cycles when idleness is disabled.

And in `m.ExitIdleMode`, we probably need a check like this at the top?
```
if m.timeout == 0 {
  m.actuallyIdle = false
  m.activeCallsCount = 0
  return
}
```

Reply:
> we still need to exit idle mode in `OnCallBegin` even when disabled

True

> We can short circuit the accounting logic

Do you think adding that short circuit would be better done as a follow-up PR since it doesn't affect correctness. If so, could we add a TODO. Or do you think it's not worth adding the short circuit?

Reply:
I believe this PR does this now.  Maybe I've missed something, in which case let me know.